# Part 1

### Q3. Import your analysis libraries, as well as your new customer data set as a dataframe.



In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Defining path
path = r'/Users/sina/Desktop/12-2022 Instacart Basket Analysis'

In [3]:
# importing customers DF
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','customers.csv'))

In [6]:
df_customers.head(5)

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Q4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.


In [7]:
# renaming columns for better readability
df_customers = df_customers.rename(columns={'First Name': 'first_name','Surnam': 'surname','Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'date_joined' :'registration_date', 'n_dependants' : 'number_dependants', 'fam_status' : 'marital_status'})

In [8]:
# checking the new titles
df_customers.head(5)

user_id first_name   surname  gender       state  age registration_date  \
0    26711    Deborah  Esquivel  Female    Missouri   48          1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36          1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35          1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40          1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26          1/1/2017   

   number_dependants marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

### Q5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [9]:
df_customers.describe()

user_id            age  number_dependants         income
count  206209.000000  206209.000000      206209.000000  206209.000000
mean   103105.000000      49.501646           1.499823   94632.852548
std     59527.555167      18.480962           1.118433   42473.786988
min         1.000000      18.000000           0.000000   25903.000000
25%     51553.000000      33.000000           0.000000   59874.000000
50%    103105.000000      49.000000           1.000000   93547.000000
75%    154657.000000      66.000000           3.000000  124244.000000
max    206209.000000      81.000000           3.000000  593901.000000

In [10]:
# Checking Data types
df_customers.dtypes

user_id               int64
first_name           object
surname              object
gender               object
state                object
age                   int64
registration_date    object
number_dependants     int64
marital_status       object
income                int64
dtype: object

In [11]:
# Checking for missing values
df_customers.isnull().sum()

user_id                  0
first_name           11259
surname                  0
gender                   0
state                    0
age                      0
registration_date        0
number_dependants        0
marital_status           0
income                   0
dtype: int64

In [12]:
df_customers[df_customers['first_name'].isnull()==True]

user_id first_name   surname  gender         state  age  \
53        76659        NaN   Gilbert    Male      Colorado   26   
73        13738        NaN     Frost  Female     Louisiana   39   
82        89996        NaN    Dawson  Female        Oregon   52   
99        96166        NaN   Oconnor    Male      Oklahoma   51   
105       29778        NaN    Dawson  Female          Utah   63   
...         ...        ...       ...     ...           ...  ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   
206044   200799        NaN  Copeland  Female        Hawaii   52   
206090   167394        NaN     Frost  Female        Hawaii   61   
206162   187532        NaN     Floyd  Female    California   39   
206171   116898        NaN   Delgado  Female      Colorado   23   

       registration_date  number_dependants marital_status  income  
53              1/1/2017                  2        married   41709  
73              1/1/2017                  0         single   82518  
82              1/1/2017                  3        married  117099  
99              1/1/2017                  1        married  155673  
105             1/1/2017                  3        married  151819  
...                  ...                ...            ...     ...  
206038         3/31/2020                  3        married   87783  
206044          4/1/2020                  2        married  108488  
206090          4/1/2020                  1        married   45275  
206162          4/1/2020                  0         single   56325  
206171          4/1/2020                  2        married   59222  

[11259 rows x 10 columns]

####  There are 11259 rows where first names are missing. since these are the first names and all the other data are intact, there is nothing needed to be done. 


In [14]:
# Checking for mixed data types 
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


#### first_name has mixed data type

In [15]:
# Convert first_name data type to string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [16]:
# Checking for mixed data types 
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

#### No more mixed data types

In [18]:
# Checking for duplicates
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, registration_date, number_dependants, marital_status, income]
Index: []

#### No duplicates found

### Q6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)


In [19]:
 # Import orders_products_merged pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated3.pkl'))


In [20]:
ords_prods_merge.shape

(32404859, 23)

In [21]:
# Merging the data
ords_prods_custs = ords_prods_merge.merge(df_customers, on = 'user_id', indicator = True)

In [22]:
ords_prods_custs.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  first_name  surname  gender    state age  \
0         Soda  ...       Linda   Nguyen  Female  Alabama  31   
1         Soda  ...       Linda   Nguyen  Female  Alabama  31   
2         Soda  ...       Linda   Nguyen  Female  Alabama  31   
3         Soda  ...       Linda   Nguyen  Female  Alabama  31   
4         Soda  ...       Linda   Nguyen  Female  Alabama  31   

  registration_date number_dependants  marital_status income  _merge  
0         2/17/2019                 3         married  40423    both  
1         2/17/2019                 3         married  40423    both  
2         2/17/2019                 3         married  40423    both  
3         2/17/2019                 3         married  40423    both  
4         2/17/2019                 3         married  40423    both  

[5 rows x 33 columns]

In [23]:
ords_prods_custs.shape

(32404859, 33)

### Q8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [24]:
# exorting file as a pickle
ords_prods_custs.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers.pkl'))
